op-guide: add upgrade FAQ (#1140)
pingcap/docs-cn: op-guide/upgrade-FAQ.md

1. 字符集（charset）问题1: unsupported modify column charset utf8mb4 not match origin utf8

    TiDB 在 v2.1.0 以及之前（包括v2.0 所有版本）默认字符集是 utf8，从 v2.1.1 开始，默认字符集变更为 utf8mb4。如果在 v2.1.0 之前建表时显式指定了 table 的 charset 为 utf8，那后升级到 v2.1.1 之后，执行 DDL 变更可能会失败。示例如下：

    升级前：v2.1.0 以及之前

    ```SQL
    tidb > create table t(a varchar(10)) charset=utf8;
    Query OK, 0 rows affected
    Time: 0.106s
    tidb > show create table t
    +-------+-------------------------------------------------------+
    | Table | Create Table                                          |
    +-------+-------------------------------------------------------+
    | t     | CREATE TABLE `t` (                                    |
    |       |   `a` varchar(10) DEFAULT NULL                        |
    |       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
    +-------+-------------------------------------------------------+
    1 row in set
    Time: 0.006s
    ```

    升级后： v2.1.1,  v2.1.2（v2.1.3 以及之后不会出现下面的问题）

    ```SQL
    tidb > alter table t change column a a varchar(20);
    (1105, u'unsupported modify column charset utf8mb4 not match origin utf8')
    ```

    解决方案：显式指定 column charset，保持和原来的 charset 一致即可。

>     ```sql

Reviewer: ```suggestion
    ```SQL
```
Some use upper and some use lower，I think we should keep unified.

    alter table t change column a a varchar(22) character set utf8;
    ```

    v2.1.3 之前不支持修改column 的charset , 所以需要指定column charset 保持和原来一致。

    v2.1.3 之前，`show create table` 不会显示 column 的 charset，即使 column 的 charset 和 table 的 charset 不一样。只能通过 http  api 拿table 的元信息查看 column 的 schema。然后根据 column 名字和 Charset 关键字搜索即可找到

    ```shell
    ▶ curl "http://$IP:10080/schema/test/t"
    {
        "ShardRowIDBits": 0,
        "auto_inc_id": 0,
        "charset": "utf8",
        "collate": "",
        "cols": [
            {
                ...
                "id": 1,
                "name": {
                    "L": "a",
                    "O": "a"
                },
                "offset": 0,
                "origin_default": null,
                "state": 5,
                "type": {
                    "Charset": "utf8",
                    "Collate": "utf8_bin",
                    "Decimal": 0,
                    "Elems": null,
                    "Flag": 0,
                    "Flen": 10,
                    "Tp": 15
                }
            }
        ],
        ...
    }
    ```

2. 字符集（charset）问题2 : unsupported modify charset from utf8mb4 to utf8

   升级前：v2.1.1, v2.1.2

>    ```sql

Reviewer: ```suggestion
   ```SQL
```
ditto

   |       |   `a` varchar(10) DEFAULT NULL                        |
   |       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
   +-------+-------------------------------------------------------+
   ```

   上面 `show create table` 只 show 出了 table 的charset，但其实 column 的 charset 是 utf8mb4，这可以用 http api 获取 schema 来确认。 这是一个 bug，即此处建表时 column 的charset 应该要和 table 保持一致为 utf8，这个问题在  v2.1.3 后已经修复。

   升级后：v2.1.3 以及之后

   ```sql
   tidb > show create table t
   +-------+--------------------------------------------------------------------+
   | Table | Create Table                                                       |
   +-------+--------------------------------------------------------------------+
   | t     | CREATE TABLE `t` (                                                 |
   |       |   `a` varchar(10) CHARSET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL |
   |       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin              |
   +-------+--------------------------------------------------------------------+
   1 row in set
   Time: 0.007s
   tidb > alter table t change column a a varchar(20);
   (1105, u'unsupported modify charset from utf8mb4 to utf8')
   ```

>    因为v2.1.3 之后支持修改column 和  table 的 charset 了，所以这里推荐修改table 的charset 为 utf8mb4。

Reviewer: ```suggestion
   因为 v2.1.3 之后支持修改 column 和  table 的 charset 了，所以这里推荐修改 table 的 charset 为 utf8mb4。
```

>    ```

Reviewer: ```suggestion
   ```SQL
```

   alter table t convert to character set utf8mb4;
   ```

>    也可以像问题1一样指定 charset。

Reviewer: ```suggestion
   也可以像问题 1 一样指定 charset。
```

---
title: 升级后常见问题与解答(upgrade FAQ)
category: deployment
---

# 升级后常见问题与解答(upgrade FAQ)

本文列出了一些升级后的可能会遇到的问题。

> ## 做 DDL 时遇到的字符集（charset）问题

Reviewer 2: s/做/执行

> TiDB 在 v2.1.0 以及之前（包括v2.0 所有版本）默认字符集是 utf8，从 v2.1.1 开始，默认字符集变更为 utf8mb4。如果在 v2.1.0 之前建表时显式指定了 table 的 charset 为 utf8，那后升级到 v2.1.1 之后，执行 DDL 变更可能会失败。

Reviewer 2: s/`包括v2.0`/`包括 v2.0`
s/utf8/UTF-8

Author: s/utf8/UTF-8
s/utf8mb4/UTF8MB4

Reviewer 2: I think we need to modify in your entire PR.

Author: Keep `utf8` `utf8mb4` in code. Use `UTF8`, `UTF8MB4` in article.

Reviewer 2: 嗯

> 记住下面2个要点：

Reviewer: ```suggestion
记住下面 2 个要点：
```

> 记住下面 2 个要点：

Reviewer 2: s/`记住下面 2 个要点：`/`2 个重点：`？

Author: em... I prefer `要点` 😂

Reviewer: I prefer `要点` too.
But I think `s/记住/注意` is better.

> 1. 在 v2.1.3 之前，不支持修改 column 的 charset。所以需要做 DDL 时需要新 column 的 charset 和 旧 column 的保持一致。

Reviewer: ```suggestion
1. 在 v2.1.3 之前，不支持修改 column 的 charset。所以执行 DDL 时需要新 column 的 charset 和旧 column 的保持一致。
```

> 2. v2.1.3 之前，`show create table` 不会显示 column 的 charset，即使 column 的 charset 和 table 的 charset 不一样。可以通过 http  api 拿table 的元信息查看 column 的 charset，下面会有示例。

Reviewer: ```suggestion
2. v2.1.3 之前，`show create table` 不会显示 column 的 charset，即使 column 的 charset 和 table 的 charset 不一样。可以通过 http  api 拿 table 的元信息查看 column 的 charset，下面会有示例。
```

> 2. v2.1.3 之前，`show create table` 不会显示 column 的 charset，即使 column 的 charset 和 table 的 charset 不一样。可以通过 http  api 拿 table 的元信息查看 column 的 charset，下面会有示例。

Reviewer: ```suggestion
2. v2.1.3 之前，即使 column 的 charset 和 table 的 charset 不一样，`show create table` 也不会显示 column 的 charset，但可以通过 http  api 拿 table 的元信息查看 column 的 charset，下面会有示例。
```

### 问题1: unsupported modify column charset utf8mb4 not match origin utf8

升级前：v2.1.0 以及之前

```SQL
tidb > create table t(a varchar(10)) charset=utf8;
Query OK, 0 rows affected
Time: 0.106s
tidb > show create table t
+-------+-------------------------------------------------------+
| Table | Create Table                                          |
+-------+-------------------------------------------------------+
| t     | CREATE TABLE `t` (                                    |
|       |   `a` varchar(10) DEFAULT NULL                        |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+-------------------------------------------------------+
1 row in set
Time: 0.006s
```

升级后： v2.1.1,  v2.1.2（v2.1.3 以及之后不会出现下面的问题）

```SQL
tidb > alter table t change column a a varchar(20);

> (1105, u'unsupported modify column charset utf8mb4 not match origin utf8')

Reviewer: `u'unsupported` 这个是什么情况？看着有点怪怪的。

Author: 用的 mycli 客户端，我改成用 mysql 吧

```

解决方案：显式指定 column charset，保持和原来的 charset 一致即可。

```SQL
alter table t change column a a varchar(22) character set utf8;
```

> 根据要点1 , 此处如果不指定column 的charset，会用默认的 utf8mb4 ，所以需要指定column charset 保持和原来一致。

Reviewer: ```suggestion
根据要点1 , 此处如果不指定 column 的 charset，会用默认的 utf8mb4 ，所以需要指定 column charset 保持和原来一致。
```

> 根据要点1 , 此处如果不指定 column 的charset，会用默认的 UTF8MB4 ，所以需要指定 column charset 保持和原来一致。

Reviewer 2: s/`根据要点1 `/`根据要点 1 `

> 根据要点2，用 http api 获取 table 元信息，然后根据 column 名字和 Charset 关键字搜索即可找到 column 的 charset。

Reviewer: ```suggestion
根据要点2，用 http api 获取 table 元信息，然后根据 column 名字和 charset 关键字搜索即可找到 column 的 charset。
```
这里的大写是预期的吗？

Author: 是，下面元信息里面是大写

> 根据要点 2，用 http api 获取 table 元信息，然后根据 column 名字和 Charset 关键字搜索即可找到 column 的 charset。

Reviewer: ```suggestion
根据要点 2，用 http api 获取 table 元信息，然后根据 column 名字和 charset 关键字搜索即可找到 column 的 charset。
```

Author: http api 返回结果中是 `Charset`

Reviewer: got it.

```shell

> ▶ curl "http://$IP:10080/schema/test/t" | python -m json.tool  #这里用了 python 的格式化 json的工具，也可以不加，此处只是为了方便注释。

Reviewer: ```suggestion
▶ curl "http://$IP:10080/schema/test/t" | python -m json.tool  # 这里用了 python 的格式化 json的工具，也可以不加，此处只是为了方便注释。
```

{
    "ShardRowIDBits": 0,
    "auto_inc_id": 0,
    "charset": "utf8",
    "collate": "",
    "cols": [
        {
            ...
            "id": 1,
            "name": {
                "L": "a",
                "O": "a"
            },
            "offset": 0,
            "origin_default": null,
            "state": 5,
            "type": {
                "Charset": "utf8",
                "Collate": "utf8_bin",
                "Decimal": 0,
                "Elems": null,
                "Flag": 0,
                "Flen": 10,
                "Tp": 15
            }
        }
    ],
    ...
}
```

> ### 问题2 : unsupported modify charset from utf8mb4 to utf8

Reviewer 3: ### 问题3 : ERROR 1366 (HY000): incorrect utf8 value f09f8c80(🌀) for column a
升级前没有对插入数据是 4-byte UTF-8 unicode encoding 检查，升级后作了检查。
```
mysql> drop table if exists t2;
Query OK, 0 rows affected (0.00 sec)

mysql> create table t2(a varchar(100) charset utf8);
Query OK, 0 rows affected (0.01 sec)

mysql> insert t2 values (unhex('f09f8c80'));
ERROR 1366 (HY000): incorrect utf8 value f09f8c80(🌀) for column a
mysql> insert t2 values (unhex('F0A48BAE'));
ERROR 1366 (HY000): incorrect utf8 value f0a48bae(𤋮) for column a
```
解决办法
* http api
```
# Enabled
curl -X POST -d "check_mb4_value_in_utf8=1" http://{TiDBIP}:10080/settings

# Disable
curl -X POST -d "check_mb4_value_in_utf8=0" http://{TiDBIP}:10080/settings
```
* session variable 
```
# Enabled
set @@session.tidb_check_mb4_value_in_utf8 = 1;

# Disable
set @@session.tidb_check_mb4_value_in_utf8 = 0;
```
* 修改配置文件
```
config.CheckMb4ValueInUtf8 = true
config.CheckMb4ValueInUtf8 = false
```

Author: @ciscoxll 可以具体到升级前后的版本吗？问题里都统一说清楚版本吧

升级前：v2.1.1, v2.1.2

```SQL
tidb > create table t(a varchar(10)) charset=utf8;
Query OK, 0 rows affected
Time: 0.109s
tidb > show create table t
+-------+-------------------------------------------------------+
| Table | Create Table                                          |
+-------+-------------------------------------------------------+
| t     | CREATE TABLE `t` (                                    |
|       |   `a` varchar(10) DEFAULT NULL                        |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+-------------------------------------------------------+
```

> 上面 `show create table` 只 show 出了 table 的charset，但其实 column 的 charset 是 utf8mb4，这可以用 http api 获取 schema 来确认。 这是一个 bug，即此处建表时 column 的charset 应该要和 table 保持一致为 utf8，这个问题在  v2.1.3 后已经修复。

Reviewer: ```suggestion
上面 `show create table` 只 show 出了 table 的 charset，但其实 column 的 charset 是 utf8mb4，这可以用 http api 获取 schema 来确认。 这是一个 bug，即此处建表时 column 的 charset 应该要和 table 保持一致为 utf8，这个问题在  v2.1.3 后已经修复。
```

升级后：v2.1.3 以及之后

```SQL
tidb > show create table t
+-------+--------------------------------------------------------------------+
| Table | Create Table                                                       |
+-------+--------------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                                 |
|       |   `a` varchar(10) CHARSET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin              |
+-------+--------------------------------------------------------------------+
1 row in set
Time: 0.007s
tidb > alter table t change column a a varchar(20);
ERROR 1105 (HY000): unsupported modify charset from utf8mb4 to utf8
```

> 因为 v2.1.3 之后支持修改 column 和  table 的 charset 了，所以这里推荐修改 table 的 charset 为 UTF8MB4。

Reviewer: ```suggestion
因为 v2.1.3 之后支持修改 column 和 table 的 charset 了，所以这里推荐修改 table 的 charset 为 UTF8MB4。
```

---

> title: 升级后常见问题与解答(upgrade FAQ)

Reviewer 4: ```suggestion
title: 升级后常见问题与解答
```

Reviewer 4: Or `升级后常见问题与解答（升级 FAQ）`

category: deployment
---

> # 升级后常见问题与解答(upgrade FAQ)

Reviewer 4: ```suggestion
# 升级后常见问题与解答
```

> 本文列出了一些升级后的可能会遇到的问题。

Reviewer 4: ```suggestion
本文列出了一些升级后可能会遇到的问题与解决办法。
```

> ## 执行 DDL 时遇到的字符集（charset）问题

Reviewer 4: ```suggestion
## 执行 DDL 操作时遇到的字符集 (charset) 问题
```

> TiDB 在 v2.1.0 以及之前（包括 v2.0 所有版本）默认字符集是 UTF8，从 v2.1.1 开始，默认字符集变更为 UTF8MB4。如果在 v2.1.0 之前建表时显式指定了 table 的 charset 为 UTF8，那后升级到 v2.1.1 之后，执行 DDL 变更可能会失败。

Reviewer 4: ```suggestion
TiDB 在 v2.1.0 以及之前版本（包括 v2.0 所有版本）中，默认字符集是 UTF8。从 v2.1.1 开始，默认字符集变更为 UTF8MB4。如果在 v2.1.0 及之前版本中，建表时显式指定了 table 的 charset 为 UTF8，那么升级到 v2.1.1 之后，执行 DDL 操作可能会失败。
```

> 注意下面 2 个要点：

Reviewer 4: ```suggestion
要避免该问题，需注意以下两个要点：
```

> 1. 在 v2.1.3 之前，不支持修改 column 的 charset。所以执行 DDL 时需要新 column 的 charset 和旧 column 的保持一致。

Reviewer 4: ```suggestion
1. 在 v2.1.3 之前，TiDB 不支持修改 column 的 charset。所以，执行 DDL 操作时，新 column 的 charset 需要和旧 column 的 charset 保持一致。
```

> 2. v2.1.3 之前，即使 column 的 charset 和 table 的 charset 不一样，`show create table` 也不会显示 column 的 charset，但可以通过 http  api 拿 table 的元信息查看 column 的 charset，下面会有示例。

Reviewer 4: ```suggestion
2. 在 v2.1.3 之前，即使 column 的 charset 和 table 的 charset 不一样，`show create table` 也不会显示 column 的 charset，但可以通过 HTTP API 获取 table 的元信息来查看 column 的 charset，下文提供了示例。
```

> ### 问题1: unsupported modify column charset utf8mb4 not match origin utf8

Reviewer 4: ```suggestion
### 问题 1：`unsupported modify column charset utf8mb4 not match origin utf8`
```

> 升级前：v2.1.0 以及之前

Reviewer 4: ```suggestion
升级前：v2.1.0 及之前版本
```

```SQL
tidb > create table t(a varchar(10)) charset=utf8;
Query OK, 0 rows affected
Time: 0.106s
tidb > show create table t
+-------+-------------------------------------------------------+
| Table | Create Table                                          |
+-------+-------------------------------------------------------+
| t     | CREATE TABLE `t` (                                    |
|       |   `a` varchar(10) DEFAULT NULL                        |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+-------------------------------------------------------+
1 row in set
Time: 0.006s
```

> 升级后： v2.1.1,  v2.1.2（v2.1.3 以及之后不会出现下面的问题）

Reviewer 5: 升级后： v2.1.1,  v2.1.2 会出现下面的问题，v2.1.3 以及之后不会出现下面的问题

> 升级后：v2.1.1, v2.1.2 会出现下面的问题，v2.1.3 以及之后不会出现下面的问题

Reviewer 4: ```suggestion
升级后：v2.1.1、v2.1.2 会出现下面的问题，v2.1.3 以及之后版本不会出现下面的问题。
```

```SQL
tidb > alter table t change column a a varchar(20);
ERROR 1105 (HY000): unsupported modify column charset utf8mb4 not match origin utf8
```

解决方案：显式指定 column charset，保持和原来的 charset 一致即可。

```SQL
alter table t change column a a varchar(22) character set utf8;
```

> 根据要点 1 , 此处如果不指定 column 的charset，会用默认的 UTF8MB4 ，所以需要指定 column charset 保持和原来一致。

Reviewer 4: ```suggestion
根据要点 1，此处如果不指定 column 的 charset，会用默认的 UTF8MB4，所以需要指定 column charset 保持和原来一致。
```

> 根据要点 2，用 http api 获取 table 元信息，然后根据 column 名字和 Charset 关键字搜索即可找到 column 的 charset。

Reviewer 4: ```suggestion
根据要点 2，用 HTTP API 获取 table 元信息，然后根据 column 名字和 Charset 关键字搜索即可找到 column 的 charset。
```

```shell
▶ curl "http://$IP:10080/schema/test/t" | python -m json.tool  # 这里用了 python 的格式化 json的工具，也可以不加，此处只是为了方便注释。
{
    "ShardRowIDBits": 0,
    "auto_inc_id": 0,
    "charset": "utf8",   # 这是 table 的 charset
    "collate": "",
    "cols": [			 # 从这里开始列举 column 的相关信息
        {
            ...
            "id": 1,
            "name": {
                "L": "a",
                "O": "a"   # column 的名字
            },
            "offset": 0,
            "origin_default": null,
            "state": 5,
            "type": {
                "Charset": "utf8",   # column a 的 charset
                "Collate": "utf8_bin",
                "Decimal": 0,
                "Elems": null,
                "Flag": 0,
                "Flen": 10,
                "Tp": 15
            }
        }
    ],
    ...
}
```

> ### 问题2 : unsupported modify charset from utf8mb4 to utf8

Reviewer 4: ```suggestion
### 问题 2：`unsupported modify charset from utf8mb4 to utf8`
```

> 升级前：v2.1.1, v2.1.2

Reviewer 4: ```suggestion
升级前：v2.1.1，v2.1.2
```

```SQL
tidb > create table t(a varchar(10)) charset=utf8;
Query OK, 0 rows affected
Time: 0.109s
tidb > show create table t
+-------+-------------------------------------------------------+
| Table | Create Table                                          |
+-------+-------------------------------------------------------+
| t     | CREATE TABLE `t` (                                    |
|       |   `a` varchar(10) DEFAULT NULL                        |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin |
+-------+-------------------------------------------------------+
```

> 上面 `show create table` 只 show 出了 table 的 charset，但其实 column 的 charset 是 UTF8MB4，这可以用 http api 获取 schema 来确认。 这是一个 bug，即此处建表时 column 的 charset 应该要和 table 保持一致为 UTF8，这个问题在  v2.1.3 后已经修复。

Reviewer 4: ```suggestion
上面 `show create table` 只显示出了 table 的 charset，但其实 column 的 charset 是 UTF8MB4，这可以通过 HTTP API 获取 schema 来确认。这是一个 bug，即此处建表时 column 的 charset 应该要和 table 保持一致为 UTF8，该问题在 v2.1.3 中已经修复。
```

> 升级后：v2.1.3 以及之后

Reviewer 4: ```suggestion
升级后：v2.1.3 及之后版本
```

```SQL
tidb > show create table t
+-------+--------------------------------------------------------------------+
| Table | Create Table                                                       |
+-------+--------------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                                 |
|       |   `a` varchar(10) CHARSET utf8mb4 COLLATE utf8mb4_bin DEFAULT NULL |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin              |
+-------+--------------------------------------------------------------------+
1 row in set
Time: 0.007s
tidb > alter table t change column a a varchar(20);
ERROR 1105 (HY000): unsupported modify charset from utf8mb4 to utf8
```

> 因为 v2.1.3 之后支持修改 column 和 table 的 charset 了，所以这里推荐修改 table 的 charset 为 UTF8MB4。

Reviewer 4: ```suggestion
因为在 v2.1.3 之后，TiDB 支持修改 column 和 table 的 charset，所以这里推荐修改 table 的 charset 为 UTF8MB4。
```

```SQL
alter table t convert to character set utf8mb4;
```

> 也可以像问题 1 一样指定 column 的 charset，保持和 column 原来的 charset  UTF8MB4 一致即可。

Reviewer 4: ```suggestion
也可以像问题 1 一样指定 column 的 charset，保持和 column 原来的 charset (UTF8MB4) 一致即可。
```

```SQL
alter table t change column a a varchar(20) character set utf8mb4;
```

> ### 问题3 : ERROR 1366 (HY000): incorrect utf8 value f09f8c80(🌀) for column a

Reviewer 4: ```suggestion
### 问题 3：`ERROR 1366 (HY000): incorrect utf8 value f09f8c80(🌀) for column a`
```

> TiDB 在 v2.1.1 以及之前，如果 charset 是 UTF8, 是没有对插入数据是 4-byte UTF8 unicode encoding 检查，在v2.1.2 以及之后，添加了检查。

Reviewer 4: ```suggestion
TiDB 在 v2.1.1 及之前版本中，如果 charset 是 UTF8，没有对 4-byte 的插入数据进行 UTF8 Unicode encoding 检查。在v2.1.2 及之后版本中，添加了该检查。
```

Reviewer 4: Please check whether this is what you mean. The original text is a little bit confusing.

> 升级前：v2.1.1 以及之前

Reviewer 4: ```suggestion
升级前：v2.1.1 及之前版本
```

```SQL
tidb> create table t(a varchar(100) charset utf8);
Query OK, 0 rows affected
tidb> insert t values (unhex('f09f8c80'));
Query OK, 1 row affected
```

> 升级后：v2.1.2 以及之后

Reviewer 4: ```suggestion
升级后：v2.1.2 及之后版本
```

```SQL
tidb> insert t values (unhex('f09f8c80'));
ERROR 1366 (HY000): incorrect utf8 value f09f8c80(🌀) for column a
```

解决方案：

> v2.1.2，该版本不支持修改 column charset, 所以只能跳过 UTF8 的检查

Reviewer 4: ```suggestion
- v2.1.2 版本：该版本不支持修改 column charset，所以只能跳过 UTF8 的检查。
```

> ```SQL

Reviewer 4: Please indent this code block with 4 spaces.

tidb > set @@session.tidb_skip_utf8_check=1;
Query OK, 0 rows affected
tidb > insert t values (unhex('f09f8c80'));
Query OK, 1 row affected
```

> v2.1.3 以及之后，建议修改 column 的 charset 为 UTF8MB4。或者也可以设置 `tidb_skip_utf8_check` 变量跳过 UTF8 的检查。

Reviewer 5: 这里补充一下如果跳过 utf8 变量的检查，在需要 TiDB 同步数据回 MySQL 的时候，可能会失败，MySQL 会做这个检查

> v2.1.3 以及之后，建议修改 column 的 charset 为 UTF8MB4。或者也可以设置 `tidb_skip_utf8_check` 变量跳过 UTF8 的检查，如果跳过 UTF8 的检查，在需要 TiDB 同步数据回 MySQL 的时候，可能会失败，因为 MySQL 会做这个检查。

Reviewer 4: ```suggestion
- v2.1.3 及之后版本：建议修改 column 的 charset 为 UTF8MB4。或者也可以设置 `tidb_skip_utf8_check` 变量跳过 UTF8 的检查。如果跳过 UTF8 的检查，在需要将数据从 TiDB 同步回 MySQL 的时候，可能会失败，因为 MySQL 会执行该检查。
```

> ```SQL

Reviewer 4: Also indent this code block with 4 spaces.

tidb > alter table t change column a a varchar(100) character set utf8mb4;
Query OK, 0 rows affected
tidb > insert t values (unhex('f09f8c80'));
Query OK, 1 row affected
```

> 关于 `tidb_skip_utf8_check` 变量，具体来说是指跳过 UTF8 和 UTF8MB4 类型对数据的合法性检查，如果跳过这个检查，在需要 TiDB 同步数据回 MySQL 的时候，可能会失败，因为 MySQL 会做这个检查。如果只想跳过 UTF8 类型的检查，可以设置 `tidb_check_mb4_value_in_utf8` 变量。

Reviewer 4: ```suggestion
关于 `tidb_skip_utf8_check` 变量，具体来说是指跳过 UTF8 和 UTF8MB4 类型对数据的合法性检查。如果跳过这个检查，在需要将数据从 TiDB 同步回 MySQL 的时候，可能会失败，因为 MySQL 执行该检查。如果只想跳过 UTF8 类型的检查，可以设置 `tidb_check_mb4_value_in_utf8` 变量。
```

`tidb_check_mb4_value_in_utf8` 在 v2.1.3 版本加入 `config.toml` 文件，可以修改配置文件里面的 `check-mb4-value-in-utf8` 后重启集群生效。

> `tidb_check_mb4_value_in_utf8` 在 v2.1.5 版本开始可以用 http api 来设置，也可以用 session 变量来设置。

Reviewer 4: ```suggestion
`tidb_check_mb4_value_in_utf8` 在 v2.1.5 版本开始可以用 HTTP API 来设置，也可以用 session 变量来设置。
```

> * http api（http api 只在单台服务器上生效）

Reviewer 4: ```suggestion
* HTTP API（HTTP API 只在单台服务器上生效）
```

> ```shell

Reviewer 4: Indentation

# Enabled
curl -X POST -d "check_mb4_value_in_utf8=1" http://{TiDBIP}:10080/settings

# Disable
curl -X POST -d "check_mb4_value_in_utf8=0" http://{TiDBIP}:10080/settings
```

> * session variable 

Reviewer 4: ```suggestion
* Session 变量
```

> ```SQL

Reviewer 4: Indentation
